src/server/transaction.h

```diff
@@ -397,9 +398,11 @@ class Transaction {
   struct MultiData {
     MultiRole role;
     MultiMode mode;
-
     std::optional<IntentLock::Mode> lock_mode;
-    absl::flat_hash_set<std::string> locks;
+
+    // Unique normalized keys used for scheduling the multi transaction.
+    std::vector<std::string> frozen_keys;
+    absl::flat_hash_set<std::string_view> frozen_keys_set;  // point to frozen_keys
 
```

Comment:
All the PAIN (😄) comes from the fact that we use CmdArgList for regular commands and we convert it to string_views only when we separate arguments by shards because we want to defer container allocations

Because of this our initialization logic is constrained with MutableSlices, so we need to provide those

EVEN though we now they all turn into string_views at the end and none of them are modified

so TODO: refactor initialization to use string_views in a clever way from the start or use variant<CmdArgList, ArgSlice>

The even greater absurdity is that we drag those mutable pointers around only to use it later for calling ToUpper() on some 10 letter argument 😵. If more of the code used the parser, we could switch transparently.

But for now let's just store it...

Reply:
it's a great point. I agree it was an overkill. But why mutability adds complexity? because you need slices of different types?

Reply:
> it's a great point. I agree it was an overkill. But why mutability adds complexity? because you need slices of different types?

It's not that mutability adds complexity, it's mostly converting spans of different types adds complexity

Mutability just is less flexible because we can't
- Use containers that rely on them as keys (like sets)
- Can't use static values (for example in StoredCmd we copy the command to a temporary buffer because we can't hand it a const char*)

Reply:
But I think we can simplify initialization code without getting rid of mutability because that's to difficult with all the code so far